### PR TITLE
Kms 27368 parallel chunks and uploads

### PR DESCRIPTION
**Pull request by atarsh**
# changes to typescript client: 
use parallel chunks upload by default.
to opt-out pass `parallelUploadsDisabled: true` in client options.
control max number of parallel calls to uploadToken.upload by passing `maxConcurrentUploadConnections` in client options. default is 6.

### CODE REVIEW

File: sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+     */
+    chunksUploaded: number;
+
+    /**
+     * number of the next chunk to start uploading
+     */
+    nextChunkIndex: number;
+}
+
+export class KalturaParallelUploadRequestAdapter extends KalturaUploadRequestAdapter {
```
Comment (talbone) [inline sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts:39]: please add top level comments

File: sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+    constructor(clientOptions: KalturaClientOptions, defaultRequestOptions: KalturaRequestOptions) {
+        super(clientOptions, defaultRequestOptions);
+        if (clientOptions.parallelUploadsDisabled) {
+            return;
+        }
+        if (KalturaUploadConnectionsManager.getTotalConnections() === -1) {
```
Comment (talbone) [inline sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts:46]: i take it -1 is the initial value indicating no value yet?
worth mentioning it a comment above.

File: sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+                }
+                else if (waitIfNoConnections) {
+                    KalturaUploadConnectionsManager.addAvailableConnectionsCallback(handleAvailableConnectionNotification);
+                    hasListenerRegistered = true;
+                }
+                return false;
+            };
+
+            const handleChunkUploadError = reason => {
+                activeAction = null;
```
Comment (talbone) [inline sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts:85]: you do not release the connection?
where is the retry initiated?

File: sources/typescript/src/adapters/kaltura-sequence-upload-request-adapter.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+import { KalturaUploadRequestAdapter } from "./kaltura-upload-request-adapter";
+
+interface UploadBySequentialChunksData {
+    enabled: boolean;
+    resume: boolean;
+    resumeAt: number;
+    finalChunk: boolean;
+}
+
+export class KalturaSequenceUploadRequestAdapter extends KalturaUploadRequestAdapter {
```
Comment (talbone) [inline sources/typescript/src/adapters/kaltura-sequence-upload-request-adapter.ts:16]: please add top level comments

File: sources/typescript/src/adapters/kaltura-sequence-upload-request-adapter.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                parameters.finalChunk = uploadChunkData.finalChunk;
+            } else {
+                console.log(`chunk upload not supported by browser or by request. Uploading the file as-is`);
+            }
+
+            // const { service, action, ks, ...queryparams } = parameters;
+            const {service, action, ...queryparams} = parameters;
+            const endpointUrl = createEndpoint(request, this.clientOptions, service, action, queryparams);
+
+            // TODO should uncomment this part and verify that upload sends the ks as part of the body and not in the query param. see https://kaltura.atlassian.net/browse/KMCNG-2401
```
Comment (talbone) [inline sources/typescript/src/adapters/kaltura-sequence-upload-request-adapter.ts:100]: ??
i didnt see one in the parallel verison.

**Review by talbone — CHANGES_REQUESTED** [review]

File: sources/typescript/src/adapters/kaltura-sequence-upload-request-adapter.ts (hunk 5 above)
Comment (atarsh) [reply sources/typescript/src/adapters/kaltura-sequence-upload-request-adapter.ts:100]: this was there before me, I just copied that to a different file :) no idea what it means.

File: sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts (hunk 2 above)
Comment (gonenradai) [inline sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts:46]: why not replace this with a more specific check like `KalturaUploadConnectionsManager.isInitialized`?
or even, just call to `initialize` which will set totalConnections internally?

File: sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts (hunk 3 above)
Comment (atarsh) [reply sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts:88]: added release connection.
no retries, same as current behavior. we discussed that and decided we will go without for now.

File: sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                            break;
+                        }
+                    }
+                }
+            };
+
+            const handleAvailableConnectionNotification = () => {
+                hasListenerRegistered = false;
+                if (KalturaUploadConnectionsManager.retrieveConnection()) {
+                    activeAction = this._uploadChunk(request, data, data.nextChunkIndex)
```
[the remark this answers is not in the document]
Comment (gonenradai) [reply sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts:128]: ignore. looked at the manager. when the adapter is notified, it is also being removed from callback queue.

File: sources/typescript/src/adapters/kaltura-upload-connections-manager.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
+export class KalturaUploadConnectionsManager {
+    private static _availableConnections: number;
+    private static _callbacks: (() => void)[];
+    private static _totalConnections: number = -1; // -1 as in not initialized
```
Comment (gonenradai) [inline sources/typescript/src/adapters/kaltura-upload-connections-manager.ts:4]: the -1 is not needed anymore as the initialized state is managed in a dedicated flag

File: sources/typescript/src/adapters/kaltura-upload-connections-manager.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * return a connection to the pool
+     */
+    public static releaseConnection(): void {
+        this._availableConnections += 1;
```
Comment (gonenradai) [inline sources/typescript/src/adapters/kaltura-upload-connections-manager.ts:38]: worth checking that there's no case where available connections > totalConnections.
Looking at the current code, the total is only set during init, but there's no verification that we're not passing it.
in fact, currently totalConnections is not used - so if it is not needed, better remove it from the code.

Comment (atarsh) [reply sources/typescript/src/adapters/kaltura-upload-connections-manager.ts:38]: I will keep it and add a log if we are passing it, which I hope we will never see..

Comment (gonenradai) [reply sources/typescript/src/adapters/kaltura-upload-connections-manager.ts:38]: client side log - surely enough if it happens to a customer we won't see it...
why not add a condition (that might never be met) like:
```
if(this._availableConnections + 1 <= this._totalConnections) {
  this._availableConnections +=1;
}
```
?

**Review by gonenradai — APPROVED** [review]
Looks OK to me

File: sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+import { KalturaClientOptions } from "../kaltura-client-options";
+import { CancelableAction } from "../cancelable-action";
+import { KalturaUploadConnectionsManager } from "./kaltura-upload-connections-manager";
+import { KalturaUploadRequestAdapter } from "./kaltura-upload-request-adapter";
+
+interface UploadByParallelChunksData {
+    chunkUploadEnabled: boolean;
+
+    /**
+     * bytes uploaded so far
```
Comment (talbone) [inline sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts:14]: what, no @var ?

**Review by talbone — APPROVED** [review]

File: sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts (hunk 9 above)
Comment (atarsh) [reply sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts:14]: never :) that's not the case...

**Pull request merged**